# Stack Overflow Developer Survey EDA

Stack Overflow is a popular website for programmers to ask questions about their code and recieve responses from other programmers. They do a yearly developer survey to better understand their community. What follows will be an exploration into the dataset in an attempt to uncover patterns among its community.

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)

## First Look

In [11]:
dtype_spec = {
    'NEWJobHunt': 'str',
    'NEWJobHuntResearch': 'str',
    'NEWLearn': 'str'
}

dev = pd.read_csv('developer_dataset.csv', dtype=dtype_spec)
dev = dev.rename(columns={
    'NEWJobHunt': 'JobHunt',
    'NEWJobHuntResearch': 'JobHuntResearch',
    'NEWLearn': 'Learn'
})
dev.head()

RespondentID  Year        Country          Employment  \
0             1  2018  United States  Employed full-time   
1             1  2019  United States  Employed full-time   
2             1  2020  United States  Employed full-time   
3             2  2018  United States  Employed full-time   
4             2  2019  United States  Employed full-time   

                                      UndergradMajor  \
0  Computer science, computer engineering, or sof...   
1  Computer science, computer engineering, or sof...   
2  Computer science, computer engineering, or sof...   
3  Computer science, computer engineering, or sof...   
4  Computer science, computer engineering, or sof...   

                                             DevType  \
0           Engineering manager;Full-stack developer   
1                              Developer, full-stack   
2                                                NaN   
3                               Full-stack developer   
4  Data or business analyst;Database administrato...   

                                  LanguageWorkedWith  \
0                                                NaN   
1                                C;C++;C#;Python;SQL   
2                                  HTML/CSS;Ruby;SQL   
3   C#;JavaScript;SQL;TypeScript;HTML;CSS;Bash/Shell   
4  Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...   

                              LanguageDesireNextYear  \
0                                                NaN   
1                                C;C#;JavaScript;SQL   
2                                    Java;Ruby;Scala   
3   C#;JavaScript;SQL;TypeScript;HTML;CSS;Bash/Shell   
4  Bash/Shell/PowerShell;HTML/CSS;JavaScript;Rust...   

                                  DatabaseWorkedWith  \
0                                                NaN   
1                                       MySQL;SQLite   
2                      MySQL;PostgreSQL;Redis;SQLite   
3  SQL Server;Microsoft Azure (Tables, CosmosDB, ...   
4                  Couchbase;DynamoDB;Firebase;MySQL   

                              DatabaseDesireNextYear  \
0                                                NaN   
1                                       MySQL;SQLite   
2                                   MySQL;PostgreSQL   
3  SQL Server;Microsoft Azure (Tables, CosmosDB, ...   
4                               Firebase;MySQL;Redis   

                                  PlatformWorkedWith  \
0                                                NaN   
1                                      Linux;Windows   
2                     AWS;Docker;Linux;MacOS;Windows   
3                                              Azure   
4  Android;AWS;Docker;IBM Cloud or Watson;iOS;Lin...   

                              PlatformDesireNextYear Hobbyist  \
0                                                NaN      NaN   
1                                      Linux;Windows       No   
2  Docker;Google Cloud Platform;Heroku;Linux;Windows      Yes   
3                                              Azure      NaN   
4  Android;AWS;Docker;IBM Cloud or Watson;Linux;S...      Yes   

                OrgSize  YearsCodePro  \
0                   NaN           NaN   
1  100 to 499 employees           1.0   
2                   NaN           8.0   
3                   NaN           4.0   
4    10 to 19 employees           8.0   

                                        JobSeek  ConvertedComp  WorkWeekHrs  \
0                                           NaN       141000.0          NaN   
1  I am not interested in new job opportunities        61000.0         80.0   
2                                           NaN            NaN          NaN   
3                                           NaN        48000.0          NaN   
4  I am not interested in new job opportunities        90000.0         40.0   

  JobHunt JobHuntResearch        Learn  
0     NaN             NaN          NaN  
1     NaN             NaN          NaN  
2     NaN             NaN  Once a year  
3     

## Column Descriptions
- **RespondentID** - A unique ID given to every respondent. As this is a yearly survey the same respondent can have multiple answers
- **Year** - The year in which responses were given
- **Country** - Country which the developer resides
- **Employment** - Employment status
- **UndergradMajor** - Undergraduate degree
- **DevType** - Developer's job title
- **LangaugeWorkedWith** - Coding languages the developer uses
- **LanguageDesireNextYear** - Coding languages the developer would like to learn
- **DatabaseWorkedWith** - Databases the developer uses
- **DatabaseDesireNextYear** - Databases the developer would like to use next year
- **PlatformsWorkedWith** - What platforms does the developer use
- **PlatformDesireNextYear** - What platforms would the developer like to use next year
- **Hobbyist** - Is coding a hobby
- **OrgSize** - How many people do they work with
- **YearsCodePro** - How many years have they coded professionally
- **JobSeek** - Are they seeking a new role
- **CovertedComp** - Salary/compensations converted to USD
- **WorkWeekHrs** - How many hours a week do they work


## Missing Data
Most questions on the survey are optional so it is likely there will be much missing data

In [17]:
dev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111209 entries, 0 to 111208
Data columns (total 21 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   RespondentID            111209 non-null  int64  
 1   Year                    111209 non-null  int64  
 2   Country                 111209 non-null  object 
 3   Employment              109425 non-null  object 
 4   UndergradMajor          98453 non-null   object 
 5   DevType                 100433 non-null  object 
 6   LanguageWorkedWith      102018 non-null  object 
 7   LanguageDesireNextYear  96044 non-null   object 
 8   DatabaseWorkedWith      85859 non-null   object 
 9   DatabaseDesireNextYear  74234 non-null   object 
 10  PlatformWorkedWith      91609 non-null   object 
 11  PlatformDesireNextYear  85376 non-null   object 
 12  Hobbyist                68352 non-null   object 
 13  OrgSize                 54804 non-null   object 
 14  YearsCodePro        

In [18]:
missing_percentage = dev.isnull().mean().round(4) * 100
missing_percentage

RespondentID               0.00
Year                       0.00
Country                    0.00
Employment                 1.60
UndergradMajor            11.47
DevType                    9.69
LanguageWorkedWith         8.26
LanguageDesireNextYear    13.64
DatabaseWorkedWith        22.79
DatabaseDesireNextYear    33.25
PlatformWorkedWith        17.62
PlatformDesireNextYear    23.23
Hobbyist                  38.54
OrgSize                   50.72
YearsCodePro              14.76
JobSeek                   45.55
ConvertedComp             17.87
WorkWeekHrs               54.06
JobHunt                   82.80
JobHuntResearch           83.20
Learn                     78.22
dtype: float64

#### Notes:
- **RespondentID, Year, Country** are only columns w/o missing data. Probably required by survey
- Percentage of missing responses ranges from 1.6% - 83.2%
- **JobHunt, JobHuntResearch, Learn** have much more missing data because they were not asked in every year. Since the percentage is so high we will drop them before further analysis

In [21]:
dev.drop(['JobHunt', 'JobHuntResearch', 'Learn'], axis=1, inplace=True)


## Summary Statistics

In [22]:
dev.describe(include='all')

RespondentID           Year        Country          Employment  \
count   111209.000000  111209.000000         111209              109425   
unique            NaN            NaN              9                   7   
top               NaN            NaN  United States  Employed full-time   
freq              NaN            NaN          53727               84707   
mean     19262.039709    2018.854832            NaN                 NaN   
std      11767.011322       0.777503            NaN                 NaN   
min          1.000000    2018.000000            NaN                 NaN   
25%       9268.000000    2018.000000            NaN                 NaN   
50%      18535.000000    2019.000000            NaN                 NaN   
75%      28347.000000    2019.000000            NaN                 NaN   
max      42857.000000    2020.000000            NaN                 NaN   

                                           UndergradMajor  \
count                                               98453   
unique                                                 19   
top     Computer science, computer engineering, or sof...   
freq                                                60852   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                      DevType          LanguageWorkedWith  \
count                  100433                      102018   
unique                  15627                       33037   
top     Developer, full-stack  C#;HTML/CSS;JavaScript;SQL   
freq                     6814                        1071   
mean                      NaN                         NaN   
std                       NaN                         NaN   
min                       NaN                         NaN   
25%                       NaN                         NaN   
50%                       NaN                         NaN   
75%                       NaN                         NaN   
max                       NaN                         NaN   

       LanguageDesireNextYear DatabaseWorkedWith DatabaseDesireNextYear  \
count                   96044              85859                  74234   
unique                  36490               7936                   8558   
top                    Python              MySQL             PostgreSQL   
freq                     1263               5850                   4474   
mean                      NaN                NaN                    NaN   
std                       NaN                NaN                    NaN   
min                       NaN                NaN                    NaN   
25%                       NaN                NaN                    NaN   
50%                       NaN                NaN                    NaN   
75%                       NaN                NaN                    NaN   
max                       NaN                NaN                    NaN   

       PlatformWorkedWith PlatformDesireNextYear Hobbyist             OrgSize  \
count               91609                  85376    68352               54804   
unique              13600                  16024        2                  10   
top               Windows                  Linux      Yes  20 to 99 employees   
freq                 4990                   4343    54733               10516   
mean                  NaN                    NaN      NaN                 NaN   
std                   NaN                    NaN      NaN                 NaN   
min                   NaN                    NaN      NaN                 NaN   
25%                   NaN                    NaN      NaN                 NaN   
50%         

#### Notes
- Some of the categorical columns have very few unique reponses while others have many, probably a result of multiple choice vs. open ended questions
- Max **WorkWeekHrs** is 475hrs, more than the number of hours in a week. This should be explored
- The middle 50% of developers have between 4 and 14 years of experience, showing the field is still growing
- Responses cam from 2018, 2019, 2020